### PR TITLE
NIFI-4805 Allow Delayed Transfer

### PR DESCRIPTION
**Pull request by patricker**
Went looking for this processor, found a work around, and found that others had already discussed building one throughout 2018 (see ticket/email thread from today, October 15th).

Happy to make changes, feedback welcome.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by bdesert** [comment]
@patricker,
this PR doesn't implement all the features requested originally in NIFI-4805, and discussed throughout the thread. Please take a look on comments from Andy, Mark and Martin.
Missing features: gradual penalization on re-try, support of EL for defining penalization period

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        final String retryAttrName = "penalization.count." + this.getIdentifier();
+        final String initialCount = flowFile.getAttribute(retryAttrName);
+        long cnt = 0;
+        if(initialCount != null) {
+            cnt = Long.parseLong(initialCount);
+        }
+
+        cnt++;
+
+        flowFile = session.putAttribute(flowFile, retryAttrName, Long.toString(cnt));
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]: as discussed under JIRA, if this processor goes for penalizing only (without re-try functionality), then I'd remove lines 77-86 at all. retry capabilities should be implemented then in separate processor.

Comment (patricker) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]: I agree. I've updated the PR. All of those other features related to retry should be handled under NIFI-3792. It sounds like there is a processor setup, he's planning to PR it soon.

**Comment by bdesert** [comment]
@patricker , the code looks good. I'll test them locally later today and will provide my feedback.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.standard;
+import org.apache.nifi.annotation.behavior.EventDriven;
+import org.apache.nifi.annotation.behavior.InputRequirement;
+import org.apache.nifi.annotation.behavior.SideEffectFree;
+import org.apache.nifi.annotation.behavior.SupportsBatching;
+import org.apache.nifi.annotation.behavior.WritesAttribute;
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]:
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.nifi.annotation.behavior.WritesAttributes;
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]:
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.components.PropertyDescriptor;
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]:
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessorInitializationContext;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]:
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Set;
+
+@EventDriven
+@SideEffectFree
+@SupportsBatching
+@Tags({"penalty", "penalize", "flowfile"})
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@CapabilityDescription("Penalizes a FlowFile.")
```
Comment (bdesert) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PenalizeFlowFile.java]:
```suggestion
@CapabilityDescription("This processor provides capability to penalize flow files. "
        + "Every flow file will be penalized as per 'Penalty Duration' property of the processor.")
```

**Comment by bdesert** [comment]
+1 LGTM
Pulled the changes, tested on local env, all looks good to go.
@markap14 , @alopresto , any additional comments?

**Comment by joewitt** [comment]
We should avoid calling this processor a 'PenalizeFlowFile' processor.  First, we have a notion of penalization in the framework already and it has a very specific meaning and it kicks in when processors reach certain failure conditions that are believed to be specific to a given flowfile or set of flowfiles and that by penalizing them the condition might improve on its own.  Second, the use for this processor in a flow might not be for the purposes of penalizing that flowfile but rather might simply be to cause a given wait or hold to occur.  I'd recommend calling it DelayFlowFile.

<...then I went and looked at the code>
It is literally just calling penalize flow file so it makes sense why the name was given.

What use case is this for again?

If we're really wanting to create a DelayFlowFile processor, which is what it sounds  like was the intent, we should do that because penalization means something else.

**Comment by patricker** [comment]
@joewitt Originally I was going to write a processor that looked at how long files had been queued and only moved them forward after the appropriate time had elapsed, a true Delay processor. But when I found this existing JIRA ticket, and since two other developers had discussed it quite a bit, I went with building a Penalize processor.

The use case is absolutely causing a delay in the Flow. Would you prefer I go back to something more like my original approach and only let FlowFile's through based on their Queued times? This way would be more configurable, and we could use an optional attribute/EL to decide how long they should be forced to delay.

**Comment by joewitt** [comment]
I do think so to be honest and i say that knowing it would be a good bit more work.  You can penalize flowfiles though that are 'not ready' - we might need to allow penalization values on a per flowfile basis (add an api call for it i mean).  I think it makes a lot more sense to the user and would be more powerful/configurable.

**Comment by patricker** [comment]
I'm confused by your message @joewitt. I understand, "Yes, make it a delay processor". But then you mention `penalize` a lot of times. My thought was if it's a delay processor then it wouldn't ever call Penalize, but just not pass FlowFile's through to the next queue until the appropriate amount of time had passed.

**Comment by joewitt** [comment]
@patricker yeah sorry - i agree.  I just meant that once you pull a flowfile from the queue and find it isn't time (and perhaps not even close to time) to send it to success you need to do something with it.  That something could be penalize it and route to a 'not ready' process so you're not constantly interrogating the same file and others can go through

**Comment by patricker** [comment]
@joewitt It's not done, more of a rough draft, but take a look at my latest commit and let me know if this is more in line with what you were thinking.

**Comment by joewitt** [comment]
closing due to inactivity

**Pull request closed without merging**